File: app/src/main/java/com/example/sl_terms/TestActivity.java
package com.example.sl_terms;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.AppCompatRadioButton;
import android.util.Base64;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import static com.example.sl_terms.BusinessLogicTest.blt;


public class TestActivity extends AppCompatActivity implements View.OnClickListener{

    Button myButton1;
    Button myButton2;

    RadioGroup rgp, rgp1;
    RadioButton rbn, rbn1;
    AvailableTest[] ID_questions;
    AvailableTest[] IdVariantVariantName;
    AvailableTest[] Picture_null;
    AvailableTest[] Picture;
    String id_studentS;
    int id_student;
    String id_testS;
    String countQuestionsS;
    String numberCorrectAnswersS;

    int id_session;



    @SuppressLint("ResourceType")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_test);
     //установить по умолчанию включенным первую  радиокнопку, чтобы нельзя было не дать ответ
     //   radioButton.setChecked(true);

    //получение данных с превой активити
        try {
            id_session=0;
            id_testS = getIntent().getStringExtra("id_test");
            id_studentS = getIntent().getStringExtra("id_student");
            int id_test = Integer.parseInt(id_testS);
            id_student = Integer.parseInt(id_studentS);

            //создание радио кнопок
            myButton1 = (Button) findViewById(R.id.button1);
            myButton2 = (Button) findViewById(R.id.button2);

            TextView textView = (TextView) findViewById(R.id.textView);


            ID_questions = blt.getIdQuestions(id_test);
            Picture_null = blt.getPictureNull(ID_questions[0].id);
            String result = Integer.toString(Picture_null[0].id);

            int countQuestions = ID_questions.length;
            countQuestionsS = Integer.toString(countQuestions);
            textView.setText(ID_questions[0].name);
            IdVariantVariantName = blt.getIdVariantVariantName(ID_questions[0].id);

            int buttons = IdVariantVariantName.length;
            AppCompatRadioButton[] rb = new AppCompatRadioButton[buttons];

            RadioGroup rgp = (RadioGroup) findViewById(R.id.radioGroup);
            rgp.setOrientation(LinearLayout.VERTICAL);


            for (int i = 0; i < buttons; i++) {
                RadioButton rbn = new RadioButton(this);
                rbn.setId(IdVariantVariantName[i].id);
                rbn.setText(IdVariantVariantName[i].name);
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
                rbn.setLayoutParams(params);
                rgp.addView(rbn);
            }

            myButton1.setOnClickListener(this);
            myButton2.setOnClickListener(this);
        }
        catch (Exception e) {
            e.printStackTrace();
            Toast toast = Toast.makeText(getApplicationContext(),
                    "Нет подключения к интернету", Toast.LENGTH_LONG );
            toast.show();
            finish();

        }

    }


    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.button1:

                //перед закрытием теста отправить последний вопрос на сервер
                RadioGroup rgp2 = (RadioGroup) findViewById(R.id.radioGroup);
                int countQuestion2 = blt.nextQuestion();
                //отправляем ответ на сервер
                int variant2 = rgp2.getCheckedRadioButtonId();

                blt.answerToCurQuestion(id_student, ID_questions[countQuestion2-1].id, variant2, id_session);

                Intent intent = new Intent(TestActivity.this, ResultActivity.class);

                numberCorrectAnswersS = Integer.toString(blt.numberOfCorrectAnswers(id_student));
                intent.putExtra("numberCorrectAnswersS", numberCorrectAnswersS);
                intent.putExtra("countQuestions", countQuestionsS);
                startActivity(intent);
                break;

            case R.id.button2: //myTextView.setText("Вы нажали на 2-ю кнопку"); break;

                RadioGroup rgp1 = (RadioGroup) findViewById(R.id.radioGroup);
                int countQuestion = blt.nextQuestion();
                //отправляем ответ на сервер
                int variant = rgp1.getCheckedRadioButtonId();
                blt.answerToCurQuestion(id_student, ID_questions[countQuestion-1].id, variant, id_session);


                 //если номер вопроса равен количеству вопросов, открыть активити с результатами
                if (countQuestion==ID_questions.length){
                    Intent intent2 = new Intent(TestActivity.this, ResultActivity.class);

                    numberCorrectAnswersS = Integer.toString(blt.numberOfCorrectAnswers(id_student));
                    intent2.putExtra("numberCorrectAnswersS", numberCorrectAnswersS);
                    intent2.putExtra("countQuestions", countQuestionsS);

                    startActivity(intent2);
                    break;

                }
                //удаляем все радио кнопки
                rgp1.removeAllViews();
                //переходим к следующему вопросу

                IdVariantVariantName = blt.getIdVariantVariantName(ID_questions[countQuestion].id);
                TextView textView = (TextView) findViewById(R.id.textView);
                //проверка есть ли рисунок
                Picture_null = blt.getPictureNull(ID_questions[countQuestion].id);
                String result = Integer.toString(Picture_null[0].id);
                //если рисунок есть то
                Picture = blt.getPicture(ID_questions[countQuestion].id);
                ImageView imageView = (ImageView) findViewById(R.id.imageView);
                imageView.setImageBitmap(null);
                imageView.destroyDrawingCache();
                if (Picture_null[0].id!= 0) {

                    byte[] decodedString = Base64.decode(Picture[0].name, Base64.DEFAULT);
                    Bitmap decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);

                    imageView.setImageBitmap(decodedByte);


                }

                textView.setText(ID_questions[countQuestion].name);

                //динамическое создание радио кнопок
                int buttons = IdVariantVariantName.length;
                AppCompatRadioButton[] rb = new AppCompatRadioButton[buttons];

                rgp1.setOrientation(LinearLayout.VERTICAL);

                for (int i = 0; i < buttons; i++) {
                    RadioButton rbn1 = new RadioButton(this);
                    rbn1.setId(IdVariantVariantName[i].id);
                    rbn1.setText(IdVariantVariantName[i].name);
                    LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
                    rbn1.setLayoutParams(params);
                    rgp1.addView(rbn1);
                }

        }
    }


    @Override
    public void onStop() {
        super.onStop();
        id_session=2;

    }

    @Override
    public void onBackPressed() {
        new AlertDialog.Builder(this)
                .setMessage("Are you sure you want to exit?")
                .setCancelable(false)
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        Intent intent = new Intent(TestActivity.this, MainActivity.class);
                        startActivity(intent);
                        TestActivity.this.finish();

                    }
                })
                .setNegativeButton("No", null)
                .show();
    }

}

File: app/src/main/java/com/example/sl_terms/activities/TestActivity.kt
package com.example.sl_terms.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Base64
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatRadioButton
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.sl_terms.BusinessLogicTest
import com.example.sl_terms.ItemMoveCallback
import com.example.sl_terms.R
import com.example.sl_terms.adapters.RVAdapter
import com.example.sl_terms.models.AvailableTest
import com.example.sl_terms.models.Option
import com.example.sl_terms.models.Question
import kotlinx.android.synthetic.main.activity_test.*


class TestActivity : AppCompatActivity(), View.OnClickListener {
    private var myButton1: Button? = null
    private var myButton2: Button? = null
    private lateinit var editText: EditText
    
    private lateinit var rv: RecyclerView
    private lateinit var rv2: RecyclerView

    private lateinit var ID_questions: Array<AvailableTest>
    private lateinit var questions: Array<Question>
    private lateinit var options: Array<Option>
    private lateinit var Picture_null: Array<AvailableTest>
    private var idPicture: Int = 0
    private lateinit var Picture: Array<AvailableTest>
    private var idStudentStr: String? = null
    private var idStudent = 0
    private lateinit var idTestStr: String
    private var countQuestionsS: String? = null
    private var numberCorrectAnswersS: String? = null
    private var idSession = 0

    private val businessLogicTest: BusinessLogicTest = BusinessLogicTest()

    @SuppressLint("ResourceType")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_test)

//получение данных с предыдущей активити
        try {
            idSession = 0
            idTestStr = intent.getStringExtra("id_test") ?: "0"
            idStudentStr = intent.getStringExtra("id_student")
            val idTest = idTestStr.toInt()
            idStudent = idStudentStr?.toInt()!!
            //создание радио кнопок
            myButton1 = findViewById<View>(R.id.button1) as Button
            myButton2 = findViewById<View>(R.id.button2) as Button
            val textView = findViewById<View>(R.id.textView) as TextView
            ID_questions = businessLogicTest.getIdQuestions(idTest)
            questions = businessLogicTest.getQuestions(idTest)
            idPicture = questions.first().idPicture ?: 0
            val countQuestions = questions.size
            countQuestionsS = countQuestions.toString()
            textView.text = questions[0].text
            options = businessLogicTest.getOptions(questions[0].id)
            val buttons = options.size
//            val rb = arrayOfNulls<AppCompatRadioButton>(buttons)
            val rgp = findViewById<View>(R.id.radioGroup) as RadioGroup
            editText = text_answer
            rgp.orientation = LinearLayout.VERTICAL
            if (questions.first().type == 0 || questions.first().type == 1 || questions.first().type == 2) {
                for (i in 0 until buttons) {
                    val rbn = if (questions.first().type == 2) CheckBox(this) else RadioButton(this)
                    rbn.id = options[i].id
                    rbn.text = options[i].name
                    val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
                    rbn.layoutParams = params
                    if (i == 0 && questions.first().type != 2)
                        rbn.isChecked = true
                    rgp.addView(rbn)
                }
            }

            if (questions.first().type == 3) {
                editText.visibility = EditText.VISIBLE
            }
            if (questions.first().type == 4) {
                val layout = llm
//                layout.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 350)
                rv = recycler_view
                rv2 = recycler_view2
                val llm = LinearLayoutManager(this)
                val llm2 = LinearLayoutManager(this)
                rv.layoutManager = llm
                rv2.layoutManager = llm2
                val optionLeft = mutableListOf<Option>()
                val optionRight = mutableListOf<Option>()

                for (i in 0 until options.size) {
                    if (options[i].type == 1) optionRight.add(options[i])
                    if (options[i].type == 0) optionLeft.add(options[i])
                }

                val adapterLeft = RVAdapter(options = optionLeft)
                val adapterRight = RVAdapter(options = optionRight)
                val callback: ItemTouchHelper.Callback = ItemMoveCallback(adapterLeft)
                val callback2: ItemTouchHelper.Callback = ItemMoveCallback(adapterRight)
                val touchHelper = ItemTouchHelper(callback)
                val touchHelper2 = ItemTouchHelper(callback2)
                touchHelper.attachToRecyclerView(recycler_view)
                touchHelper2.attachToRecyclerView(recycler_view2)
                rv.adapter = adapterLeft
                rv2.adapter = adapterRight
                layout.visibility = LinearLayout.VISIBLE

            }
            myButton1!!.setOnClickListener(this)
            myButton2!!.setOnClickListener(this)
        } catch (e: Exception) {
            e.printStackTrace()
            val toast = Toast.makeText(applicationContext,
                    "Нет подключения к интернету", Toast.LENGTH_LONG)
            toast.show()
            finish()
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.button1 -> {
                endTest()
            }
            R.id.button2 -> {
                val rgp1 = findViewById<View>(R.id.radioGroup) as RadioGroup
                val countQuestion: Int = businessLogicTest.nextQuestion()
                when (questions[countQuestion - 1].type) {
                    0, 1 -> {
                        //отправляем ответ на сервер
                        val variant2 = rgp1.checkedRadioButtonId
                        businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion - 1].id, variant2, "", idSession)
                        Log.e("TAG", variant2.toString())
                    }
                    2 -> {
                        val countButtons = rgp1.childCount
                        var checkedOptions = ""
                        for (i in 0 until countButtons) {
                            val option = rgp1.getChildAt(i) as CheckBox
                            if (option.isChecked) {
                                checkedOptions += "${option.id},"
                            }
                        }
                        Log.e("TAG", checkedOptions[checkedOptions.lastIndex].toString())
                        if (checkedOptions[checkedOptions.lastIndex] == ',') checkedOptions = checkedOptions.dropLast(1)
                        businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion - 1].id, rgp1.getChildAt(0).id, checkedOptions, idSession)
                        Log.e("TAG", checkedOptions)
                    }
                    3 -> {
                        val textAnswer = text_answer.text
                        businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion - 1].id, options.first().id, textAnswer.toString().toUpperCase(), idSession)
                        Log.e("TAG", textAnswer.toString())
                    }
                    4 -> {
                        val resView = recycler_view
                        val resView2 = recycler_view2
                        val countOptions = resView.childCount
                        var otherAnswers = ""
                        for (i in 0 until countOptions) {
                            val option = resView.getChildAt(i).id
                            val option2 = resView2.getChildAt(i).id
                            otherAnswers += "${option}:$option2,"
                        }
                        if (otherAnswers[otherAnswers.lastIndex] == ',') otherAnswers = otherAnswers.dropLast(1)
                        businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion - 1].id, options.first().id, otherAnswers, idSession)
                        Log.e("TAG", otherAnswers)
                    }
                }
                //отправляем ответ на сервер
//                val variant = rgp1.checkedRadioButtonId
//                BusinessLogicTest.blt.answerToCurQuestion(idStudent, questions[countQuestion - 1].id, variant, idSession)
                //если номер вопроса равен количеству вопросов, открыть активити с результатами
                if (countQuestion == questions.size) {
                    val intent2 = Intent(this@TestActivity, ResultActivity::class.java)
                    numberCorrectAnswersS = BusinessLogicTest.businessLogicTest.numberOfCorrectAnswers(id_student = idStudent, idTest = idTestStr.toInt()).toString()
                    intent2.putExtra("numberCorrectAnswersS", numberCorrectAnswersS)
                    intent2.putExtra("countQuestions", countQuestionsS)
                    startActivity(intent2)
                }
                //удаляем все радио кнопки
                rgp1.removeAllViews()
                val layout = llm
                layout.visibility = LinearLayout.GONE
                editText.visibility = EditText.GONE
                //переходим к следующему вопросу
                options = BusinessLogicTest.businessLogicTest.getOptions(questions[countQuestion].id)
                val textView = findViewById<View>(R.id.textView) as TextView
                //проверка есть ли рисунок
                Picture_null = BusinessLogicTest.businessLogicTest.getPictureNull(questions[countQuestion].id)
//                val result = Picture_null[0].id.toString()
                //если рисунок есть то
                Picture = BusinessLogicTest.businessLogicTest.getPicture(questions[countQuestion].id)
                val imageView = findViewById<View>(R.id.imageView) as ImageView
                imageView.setImageBitmap(null)
                imageView.destroyDrawingCache()
                if (Picture_null[0].id != 0) {
                    val decodedString = Base64.decode(Picture[0].name, Base64.DEFAULT)
                    val decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.size)
                    imageView.setImageBitmap(decodedByte)
                }
                textView.text = questions[countQuestion].text
                //динамическое создание радио кнопок
                val buttons = options.size
                val rb = arrayOfNulls<AppCompatRadioButton>(buttons)
                rgp1.orientation = LinearLayout.VERTICAL
                var i = 0
                if (questions[countQuestion].type == 0 || questions[countQuestion].type == 1 || questions[countQuestion].type == 2) {
                    while (i < buttons) {
                        val rbn1 = if (questions[countQuestion].type == 2) CheckBox(this) else RadioButton(this)
                        rbn1.id = options[i].id
                        rbn1.text = options[i].name
                        val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
                        rbn1.layoutParams = params
                        if (i == 0 && questions[countQuestion].type != 2)
                            rbn1.isChecked = true
                        rgp1.addView(rbn1)
                        i++
                    }
                }
                if (questions[countQuestion].type == 3) {
                    editText.visibility = EditText.VISIBLE
                }
                if (questions[countQuestion].type == 4) {
                    val layout = llm
//                layout.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 350)
                    rv = recycler_view
                    rv2 = recycler_view2
                    val llm = LinearLayoutManager(this)
                    val llm2 = LinearLayoutManager(this)
                    rv.layoutManager = llm
                    rv2.layoutManager = llm2
                    val optionLeft = mutableListOf<Option>()
                    val optionRight = mutableListOf<Option>()

                    for (i in 0 until options.size) {
                        if (options[i].type == 1) optionRight.add(options[i])
                        if (options[i].type == 0) optionLeft.add(options[i])
                    }

                    val adapterLeft = RVAdapter(options = optionLeft)
                    val adapterRight = RVAdapter(options = optionRight)
                    val callback: ItemTouchHelper.Callback = ItemMoveCallback(adapterLeft)
                    val callback2: ItemTouchHelper.Callback = ItemMoveCallback(adapterRight)
                    val touchHelper = ItemTouchHelper(callback)
                    val touchHelper2 = ItemTouchHelper(callback2)
                    touchHelper.attachToRecyclerView(recycler_view)
                    touchHelper2.attachToRecyclerView(recycler_view2)
                    rv.adapter = adapterLeft
                    rv2.adapter = adapterRight
                    layout.visibility = LinearLayout.VISIBLE
                }

            }
        }
    }

    fun endTest() {
        val rgp2 = findViewById<View>(R.id.radioGroup) as RadioGroup
        val countQuestion2: Int = businessLogicTest.nextQuestion()
        when (questions[countQuestion2 - 1].type) {
            0, 1 -> {
                //отправляем ответ на сервер
                val variant2 = rgp2.checkedRadioButtonId
                businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion2 - 1].id, variant2, "", idSession)
            }
            2 -> {
                val countButtons = rgp2.childCount
                var checkedOptions = ""
                for (i in 0 until countButtons) {
                    val option = rgp2.getChildAt(i) as CheckBox
                    if (option.isChecked) {
                        checkedOptions += "${option.id},"
                    }
                }
                if (checkedOptions[checkedOptions.lastIndex] == ',') checkedOptions = checkedOptions.dropLast(1)
                businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion2 - 1].id, rgp2.getChildAt(0).id, checkedOptions, idSession)
            }
            3 -> {
                val textAnswer = text_answer.text
                businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion2 - 1].id, options.first().id, textAnswer.toString().toUpperCase(), idSession)
            }
            4 -> {
                val resView = recycler_view
                val resView2 = recycler_view2
                val countOptions = resView.childCount
                var otherAnswers = ""
                for (i in 0 until countOptions) {
                    val option = resView.getChildAt(i).id
                    val option2 = resView2.getChildAt(i).id
                    otherAnswers += "${option}:$option2,"
                }
                if (otherAnswers[otherAnswers.lastIndex] == ',') otherAnswers = otherAnswers.dropLast(1)
                businessLogicTest.answerToCurQuestion(idStudent, questions[countQuestion2 - 1].id, options.first().id, otherAnswers, idSession)
                Log.e("TAG", otherAnswers)
            }
        }

        val intent = Intent(this@TestActivity, ResultActivity::class.java)
        numberCorrectAnswersS = BusinessLogicTest.businessLogicTest.numberOfCorrectAnswers(idStudent, idTestStr.toInt()).toString()
        intent.putExtra("numberCorrectAnswersS", numberCorrectAnswersS)
        intent.putExtra("countQuestions", countQuestionsS)
        startActivity(intent)
        finish()
    }

    public override fun onStop() {
        super.onStop()
        idSession = 2
    }

    override fun onResume() {
        super.onResume()
        if (idSession == 2) endTest()
    }

    override fun onBackPressed() {
        AlertDialog.Builder(this)
                .setMessage("Вы действительно хотите выйти?")
                .setCancelable(false)
                .setPositiveButton("Да") { dialog, id ->
                    val intent = Intent(this@TestActivity, MainActivity::class.java)
                    startActivity(intent)
                    finish()
                }
                .setNegativeButton("Нет", null)
                .show()
    }
}